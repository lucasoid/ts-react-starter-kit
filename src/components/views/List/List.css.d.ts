export const listItem: string;
export const category: string;
export const name: string;
export const quantity: string;
export const active: string;
export const inactive: string;
