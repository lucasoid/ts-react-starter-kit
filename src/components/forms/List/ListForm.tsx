import * as React from 'react';
import uuidv4 from 'uuid/v4';
import { Form } from '~components/forms/Form';
import { TextField } from '~components/forms/Inputs';

interface IListFormProps {
    onSubmit: Function;
    onCancel: Function;
}
export const ListForm: React.FC<IListFormProps> = props => {
    const [name, setName] = React.useState('');
    const [members, setMembers] = React.useState('');

    return (
        <Form
            onSubmit={() => {
                props.onSubmit({
                    id: uuidv4(),
                    name: name,
                    owner: 'me@domain.com',
                    members: members.split(','),
                });
            }}
            onCancel={props.onCancel}
            submitText={'Create'}
            cancelText={'Cancel'}
        >
            <TextField value={name} label="List name" onChange={setName} />
            <TextField value={'me@domain.com'} label="List owner" onChange={setName} disabled={true} />
            <TextField value={members} label="Share with" onChange={setMembers} />
        </Form>
    );
};
