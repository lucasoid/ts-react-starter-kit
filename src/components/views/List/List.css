.listItem {
    padding: 0.5em;
    display: flex;
    flex-direction: row;
}

.listItem.inactive {
    text-decoration: line-through;
}

.listItem.active {
    text-decoration: none;
}

.listItem .category {
    flex: 0 0 30%;
}

.listItem .name {
    flex: 0 0 30%;
}

.listItem .quantity {
    flex: 0 0 10%;
}
