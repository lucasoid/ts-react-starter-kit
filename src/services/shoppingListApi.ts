export interface IList {
    id: string;
    name: string;
    owner: string;
    members?: string[];
}

export interface IListItem {
    id: string;
    listId: string;
    category: string;
    name: string;
    quantity: number;
    isActive: boolean;
}

const LATENCY: number = 1000; // to mock the latency of the API in ms

let sampleListData: IList[] = [
    {
        id: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        name: 'Weekly shopping list',
        owner: 'me@domain.com',
        members: ['you@domain.com'],
    },
    {
        id: '0ba0e852-5e9f-44a7-9d6e-d113b4a63173',
        name: 'Hardware supplies',
        owner: 'me@domain.com',
        members: ['him@domain.com'],
    },
    {
        id: '2a335b66-a92a-4fb0-94e7-4ec1c673937e',
        name: 'Christmas list - East coast fam',
        owner: 'you@domain.com',
        members: ['me@domain.com', 'her@domain.com'],
    },
];

let sampleListItemData: IListItem[] = [
    {
        id: '3dd08312-7cca-4467-b443-09296cb54245',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'produce',
        name: 'avocados',
        quantity: 3,
        isActive: true,
    },
    {
        id: '7e81ee74-5fc1-4395-be2b-4b4616f6ab59',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'produce',
        name: 'tomatoes',
        quantity: 1,
        isActive: true,
    },
    {
        id: '43529c83-a355-4099-b6f5-2179f210d4b9',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'produce',
        name: 'onions',
        quantity: 1,
        isActive: true,
    },
    {
        id: '36de2c94-bfc6-41dc-91f7-6fd812107785',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'produce',
        name: 'cilantro - bunch',
        quantity: 1,
        isActive: true,
    },
    {
        id: '1bf8f08a-7ddd-45b9-b401-ca8e13870382',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'spices',
        name: 'salt',
        quantity: 1,
        isActive: false,
    },
    {
        id: '293df6d2-e9d6-4f42-92f6-257543d6dde4',
        listId: '88a1051a-d69b-4655-8ba4-be8d9a44599f',
        category: 'spices',
        name: 'pepper',
        quantity: 1,
        isActive: false,
    },
    {
        id: '8253fc7c-14fd-47e7-ad08-4b560cfb34ef',
        listId: '0ba0e852-5e9f-44a7-9d6e-d113b4a63173',
        category: 'Lumber',
        name: "2x4s - 4'",
        quantity: 8,
        isActive: true,
    },
    {
        id: '7e1b906c-f6d4-467b-9cea-7072d4cb1d43',
        listId: '0ba0e852-5e9f-44a7-9d6e-d113b4a63173',
        category: 'Lighting',
        name: 'LED floodlight bulbs - 2700k',
        quantity: 2,
        isActive: true,
    },
];

let listSubscribers: Function[] = [];
let listItemSubscribers: Object = {};

export const subscribeToLists = (cb: Function) => {
    if (listSubscribers.indexOf(cb) === -1) listSubscribers.push(cb);
};

export const unsubscribeToLists = (cb: Function) => {
    let index = listSubscribers.indexOf(cb);
    if (index !== -1) {
        listSubscribers.splice(index, 1);
    }
};

export const subscribeToListItems = (listId: string, cb: Function) => {
    if (listItemSubscribers[listId] === undefined) {
        listItemSubscribers[listId] = [cb];
    } else {
        listItemSubscribers[listId] = [...listItemSubscribers[listId], cb];
    }
};

export const unsubscribeToListItems = (listId: string, cb: Function) => {
    if (listItemSubscribers[listId] !== undefined) {
        let index = listItemSubscribers[listId].indexOf(cb);
        if (index !== -1) {
            listItemSubscribers[listId].splice(index, 1);
        }
    }
};

const onUpdateLists = (data: IList[]) => {
    listSubscribers.forEach(fn => fn(data));
};

const onUpdateListItems = (listId: string, data: IListItem[]) => {
    if (listItemSubscribers[listId] !== undefined) {
        listItemSubscribers[listId].forEach(fn => fn(data));
    }
};

export const fetchLists = async () => {
    const lists: IList[] = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(sampleListData);
        }, LATENCY);
    });
    onUpdateLists(lists);
    return lists;
};

export const createList = async (list: IList) => {
    const newList: IList = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(list);
        }, LATENCY);
    });
    sampleListData.push(newList);
    onUpdateLists(sampleListData);
    return list;
};

export const editList = async (list: IList) => {
    const saved: IList = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(saved);
        }, LATENCY);
    });
    const index = sampleListData.findIndex(_list => _list.id === list.id);
    sampleListData.splice(index, 1, saved);
    onUpdateLists(sampleListData);
    return saved;
};

export const fetchListItems = async (listId: string) => {
    const items: IListItem[] = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(sampleListItemData.filter(item => item.listId === listId));
        }, LATENCY);
    });
    onUpdateListItems(listId, items);
    return items;
};

export const createListItem = async (item: IListItem) => {
    const saved: IListItem = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(item);
        }, LATENCY);
    });
    sampleListItemData.push(saved);
    onUpdateListItems(item.listId, sampleListItemData);
    return item;
};

export const editListItem = async (item: IListItem) => {
    const saved: IListItem = await new Promise((resolve, reject) => {
        window.setTimeout(() => {
            resolve(item);
        }, LATENCY);
    });
    const index = sampleListItemData.findIndex(_item => _item.id === item.id);
    sampleListItemData.splice(index, 1, saved);
    onUpdateListItems(
        item.listId,
        sampleListItemData.filter(_item => _item.listId === item.listId),
    );
    return saved;
};
