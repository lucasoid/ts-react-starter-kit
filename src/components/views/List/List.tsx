import * as React from 'react';
import { Link, /*useParams,*/ withRouter } from 'react-router-dom';
import { View } from '~components/views/View';
import { Button } from '~components/ui/Button';
import {
    fetchListItems,
    IListItem,
    subscribeToListItems,
    unsubscribeToListItems,
    IList,
    fetchLists,
} from '~services/shoppingListApi';
import * as styles from './List.css';

interface IListProps {
    match?: any;
}

interface IListState {
    list: IList;
    items: IListItem[];
}

class InertList extends React.Component<IListProps, IListState> {
    state = {
        list: {
            id: null,
            name: null,
            owner: null,
            members: [],
        },
        items: [],
    };

    subscriber = items => {
        this.setState({ items: items });
    };

    async componentDidMount() {
        const { listId } = this.props.match.params;
        let lists = await fetchLists();
        let list = lists.find(_list => _list.id === listId);
        this.setState({ list }, () => {
            subscribeToListItems(this.state.list.id, this.subscriber);
            fetchListItems(this.state.list.id);
        });
    }

    componentWillUnmount() {
        unsubscribeToListItems(this.state.list.id, this.subscriber);
    }

    render() {
        return (
            <View>
                <Link to="/">‚Üê All lists</Link>
                <h2>{this.state.list.name}</h2>
                <aside>Owner: {this.state.list.owner}</aside>
                <aside>Members: {this.state.list.members.join(', ')}</aside>
                {this.state.items.map((item, index) => (
                    <div
                        className={`${styles.listItem} ${item.isActive ? styles.active : styles.inactive}`}
                        key={item.id + index}
                    >
                        <div className={styles.category}>{item.category}</div>
                        <div className={styles.name}>{item.name}</div>
                        <div className={styles.quantity}>{item.quantity}</div>
                    </div>
                ))}
                <Button styles={{ marginTop: '1em', backgroundColor: 'teal' }}>+ New item</Button>
            </View>
        );
    }
}

export const List = withRouter(InertList);
