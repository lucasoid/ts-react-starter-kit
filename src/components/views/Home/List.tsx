import * as React from 'react';
import { Link } from 'react-router-dom';
import { Loading } from '~components/loading';
import * as styles from './Home.css';
import { ThemeContext, Themes } from '~theme';
import { IList } from '~services/shoppingListApi';

interface IListProps {
    list: IList;
}
export const List: React.FC<IListProps> = props => {
    let { theme } = React.useContext(ThemeContext);

    return (
        <div className={`${styles.menuItem} ${theme === Themes.DARK ? styles.dark : styles.light}`}>
            {!props.list ? (
                <>
                    <Loading el={'a'} lines={1} width={'35%'} />
                    <Loading el={'aside'} lines={1} width={'25%'} />
                </>
            ) : (
                <>
                    <Link to={`/lists/${props.list.id}`} title={props.list.name}>
                        {props.list.name}
                    </Link>
                    <aside>Created by {props.list.owner}</aside>
                </>
            )}
        </div>
    );
};
