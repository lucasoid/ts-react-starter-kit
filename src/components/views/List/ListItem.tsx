import * as React from 'react';
import { Button } from '~components/ui/Button';
import { Loading } from '~components/Loading';
import * as styles from './List.css';
import { ThemeContext, Themes } from '~theme';
import { IListItem } from '~services/shoppingListApi';

interface IListItemProps {
    item: IListItem;
    onClick: Function;
}
export const ListItem: React.FC<IListItemProps> = props => {
    let { item } = props;
    let { theme } = React.useContext(ThemeContext);

    return !item ? (
        <div className={`${styles.listItem} ${styles.active} ${theme === Themes.DARK ? styles.dark : styles.light}`}>
            <div className={styles.checkbox}>
                <Button plaintext={true} styles={{ outline: 'none' }}>
                    ☐
                </Button>
            </div>
            <div className={styles.name}>
                <Loading width={'100%'} lines={1} el={'span'} />
            </div>
            <div className={styles.category}>
                <Loading width={'100%'} lines={1} el={'span'} />
            </div>
            <div className={styles.quantity}>
                <Loading width={'100%'} lines={1} el={'span'} />
            </div>
        </div>
    ) : (
        <div
            className={`
                ${styles.listItem}
                ${item.isActive ? styles.active : styles.inactive}
                ${theme === Themes.DARK ? styles.dark : styles.light}
            `}
            onClick={() => props.onClick(item)}
        >
            <div className={styles.checkbox}>
                <Button plaintext={true} styles={{ outline: 'none' }}>
                    {item.isActive ? '☐' : '☑'}
                </Button>
            </div>
            <div className={styles.name}>{item.name}</div>
            <div className={styles.category}>{item.category}</div>
            <div className={styles.quantity}>{item.quantity}</div>
        </div>
    );
};
