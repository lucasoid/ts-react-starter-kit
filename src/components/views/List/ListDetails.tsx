import * as React from 'react';
import { Button, ButtonTypes } from '~components/ui/Button';
import { Loading, Await } from '~components/Loading';
import { ListItem } from './ListItem';
import { Modal, ModalWidth } from '~components/ui/Modal';
import { ListItemForm } from '~components/forms/ListItem';
import {
    fetchListItems,
    IListItem,
    subscribeToListItems,
    unsubscribeToListItems,
    IList,
    fetchLists,
    editListItem,
    createListItem,
} from '~services/shoppingListApi';
import { E404 } from '~components/errors/ErrorTypes';

interface IListDetailsProps {
    listId: string;
}

export const ListDetails: React.FC<IListDetailsProps> = props => {
    let [list, setList] = React.useState({ id: null, name: null, owner: null, members: [] } as IList);
    let [loading, setLoading] = React.useState(true);
    let [items, setItems] = React.useState([]);
    let [notFound, setNotFound] = React.useState(false);
    let [showCreateDialog, setShowCreateDialog] = React.useState(false);
    React.useEffect(() => {
        const subscriber = function(items) {
            setItems(items);
            setLoading(false);
        };
        fetchLists().then(lists => {
            let list = lists.find(_list => _list.id === props.listId);
            if (!list) {
                setNotFound(true);
            } else {
                setList(list);
                subscribeToListItems(props.listId, subscriber);
                fetchListItems(list.id);
            }
        });
        return () => unsubscribeToListItems(props.listId, subscriber);
    }, [props.listId]);

    function toggle(item: IListItem) {
        item.isActive = !item.isActive;
        // update the API, which will trigger the subscriber when it completes
        editListItem(item);
        // update state optimistically while we wait
        let _items = [...items];
        _items.splice(
            items.findIndex(_item => _item.id === item.id),
            1,
            item,
        );
        setItems(_items);
    }

    function handleCreate(item: IListItem) {
        item.listId = props.listId;
        createListItem(item);
        let _items = [...items];
        _items.push(item);
        setItems(_items);
        setShowCreateDialog(false);
    }

    if (notFound) throw new E404('List not found');

    return (
        <Await
            loaded={!loading}
            renderFallback={() => (
                <>
                    <Loading el={'h2'} width={'25%'} lines={1}></Loading>
                    <Loading el={'aside'} width={'25%'} lines={1}></Loading>
                    <Loading el={'aside'} width={'25%'} lines={1}></Loading>
                    <ListItem item={null} key={1} onClick={() => {}} />
                    <ListItem item={null} key={2} onClick={() => {}} />
                    <ListItem item={null} key={3} onClick={() => {}} />
                </>
            )}
        >
            <h2>{list.name}</h2>
            <aside>Owner: {list.owner}</aside>
            <aside>Members: {list.members.join(', ')}</aside>
            {items.map((item, index) => (
                <ListItem item={item} key={item.id} onClick={toggle} />
            ))}
            <Button onClick={() => setShowCreateDialog(true)} type={ButtonTypes.PRIMARY} styles={{ marginTop: '1em' }}>
                + New item
            </Button>
            {showCreateDialog && (
                <Modal width={ModalWidth.MEDIUM} onDismiss={() => setShowCreateDialog(false)}>
                    <ListItemForm onSubmit={handleCreate} onCancel={() => setShowCreateDialog(false)} />
                </Modal>
            )}
        </Await>
    );
};
