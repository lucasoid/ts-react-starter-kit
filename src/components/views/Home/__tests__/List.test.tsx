import * as React from 'react';
import { shallow } from 'enzyme';
import { List } from '../List';
import { IList } from '~services/shoppingListApi';
import { Themes } from '~theme';

jest.mock('react', () => {
    const original = jest.requireActual('react');
    return {
        __esModule: true,
        ...original,
        useContext: jest.fn(),
    };
});

describe('List', () => {
    it('displays a themed List item', () => {
        let list: IList = { id: '1', name: 'My cool list', owner: 'me@domain.com', members: [] };
        // @ts-ignore
        React.useContext.mockImplementationOnce(() => ({ theme: Themes.LIGHT }));
        let wrapper = shallow(<List list={list} />);
        expect(wrapper).toMatchSnapshot();

        // @ts-ignore
        React.useContext.mockImplementationOnce(() => ({ theme: Themes.DARK }));
        wrapper = shallow(<List list={list} />);
        expect(wrapper).toMatchSnapshot();
    });
});
