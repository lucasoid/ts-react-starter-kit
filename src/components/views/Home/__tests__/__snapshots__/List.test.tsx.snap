// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List displays a themed List item 1`] = `
<div
  className="menuItem light"
>
  <Link
    title="My cool list"
    to="/lists/1"
  >
    My cool list
  </Link>
  <aside>
    Created by 
    me@domain.com
  </aside>
</div>
`;

exports[`List displays a themed List item 2`] = `
<div
  className="menuItem dark"
>
  <Link
    title="My cool list"
    to="/lists/1"
  >
    My cool list
  </Link>
  <aside>
    Created by 
    me@domain.com
  </aside>
</div>
`;
